### PR TITLE
[FIXED JENKINS-14538] Move tool configuration to separate page

### PR DESCRIPTION
**Pull request by daniel-beck**
[JENKINS-14538](https://issues.jenkins-ci.org/browse/JENKINS-14538)

@jenkinsci/code-reviewers (especially @imod), @reviewbybees 

---

Work in progress:
- [x] Basic changes done
- [x] Needs better icon (or rather a different icon for _Configure Jenkins_, a large gear may work);
- [ ] Change package names
- [ ] Fix tests 

---

> screen shot 2016-02-03 at 18 03 11

---

> screen shot 2016-02-03 at 18 18 36

### CODE REVIEW

**Comment by stephenc** [comment]
(bee) on the simple move... though I would like to refactor the tools to allow for hierarchical configuration of tools when using folders... but outside the scope of this change

**Comment by oleg-nenashev** [comment]
Yes, I would also like to see hierarchical tool pages in 2.0. Especially for the big configs like Custom Tools

**Comment by uhafner** [comment]
Shouldn't we wait with such UI changes until the tabbing concept in #2006 is finished? Otherwise users need to learn a new UI two times. I think using tabs for the system configuration makes sense, too.

**Comment by daniel-beck** [comment]
@uhafner I doubt we'd use the config tabs from 2006 for the global config. I discussed this with Tom and that configuration page would likely appear pretty broken with tabs judging by its current structure.

OTOH it would likely be feasible to use tabs for the tools page.

**Comment by daniel-beck** [comment]
@oleg-nenashev @stephenc I doubt we're able to get anything more significant done in time for 2.0. Would you prefer to keep the current global configuration page structure over this change in that case?

**Comment by imod** [comment]
yeah, a hierarchy would be great - but I agree this should be implemented in an other PR. Beside that, I think this a very good start!

**Comment by oleg-nenashev** [comment]
> I doubt we're able to get anything more significant done in time for 2.0. Would you prefer to keep the current global configuration page structure over this change in that case?

Nope. I would prefer to get this change integrated. Tabbed view may be useful, of course. And I think that we will be able to change the layout after 2.0 Big Bang, because there should not be data structure changes.

**Comment by daniel-beck** [comment]
Fixed the icons: The logical approach seems to be changing regular *Configure* to use the gear icon, which frees the previously used tools icon for the tools configuration.

> screen shot

WDYT?

**Comment by stephenc** [comment]
:bee:

**Comment by aheritier** [comment]
:bee: I like

**Comment by batmat** [comment]
:+1: very good thing

2016-03-08 21:50 GMT+01:00 Arnaud Heritier <notifications@github.com>:

**Comment by jglick** [comment]
Test failures look real—tests need to be updated to match.

File: core/src/main/java/jenkins/model/GlobalConfigurationCategory.java [hunk 1]
```diff
@@ -73,11 +71,28 @@ public String getDisplayName() {
...
         }
     }
+
+    /**
+     * Global configuration of tool locations and installers.
+     *
+     * @since TODO
+     */
+    @Extension
+    public static class Tools extends GlobalConfigurationCategory {
```
Comment (jglick) [inline core/src/main/java/jenkins/model/GlobalConfigurationCategory.java]: Perhaps this would make more sense in `ToolDescriptor`?

File: core/src/main/java/jenkins/model/GlobalToolConfiguration.java [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package jenkins.model;
```
Comment (jglick) [inline core/src/main/java/jenkins/model/GlobalToolConfiguration.java]: `jenkins.tools` would perhaps be a better fit. (To match `hudson.tools`, but…well, under `jenkins.*` rather than `hudson.*`.)

File: core/src/main/java/jenkins/model/GlobalToolConfiguration.java [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+        }
+
+        return result;
+    }
+
+    private boolean configureDescriptor(StaplerRequest req, JSONObject json, Descriptor<?> d) throws Descriptor.FormException {
+        String name = d.getJsonSafeClassName();
+        JSONObject js = json.has(name) ? json.getJSONObject(name) : new JSONObject(); // if it doesn't have the property, the method returns invalid null object.
+        json.putAll(js);
+        return d.configure(req, js);
```
Comment (jglick) [inline core/src/main/java/jenkins/model/GlobalToolConfiguration.java]: This seems like rather low-level form binding code that ought to live in a more generic place.

File: core/src/main/resources/hudson/model/Label/sidepanel.jelly [hunk 4]
```diff
@@ -34,7 +34,7 @@ THE SOFTWARE.
       <l:task contextMenu="false" href="${rootURL}/" icon="icon-up icon-md" title="${%Back to Dashboard}"/>
       <l:task contextMenu="false" href="${url}" icon="icon-attribute icon-md" title="${%Overview}"/>
       <j:if test="${it.atom}">
-        <l:task href="${url}/configure" icon="icon-setting icon-md" title="${%Configure}" permission="${app.ADMINISTER}" />
+        <l:task href="${url}/configure" icon="icon-gear2 icon-md" title="${%Configure}" permission="${app.ADMINISTER}" />
```
Comment (jglick) [inline core/src/main/resources/hudson/model/Label/sidepanel.jelly]: How are these changes related to the PR?

File: core/src/main/java/jenkins/model/GlobalToolConfiguration.java (hunk 3 above)
Comment (jglick) [inline core/src/main/java/jenkins/model/GlobalToolConfiguration.java]: In fact this whole class seems like it was just copied & pasted from `GlobalSecurityConfiguration`, suggesting that there could be a common superclass that handles this boilerplate and offers the `index` view, with some placeholders for specialized behaviors.

File: core/src/main/java/hudson/Functions.java [hunk 5]
```diff
@@ -956,11 +956,13 @@ public static String getFooterURL() {
...
 
             if (d instanceof GlobalConfiguration) {
-                if (predicate.apply(((GlobalConfiguration)d).getCategory()))
+                if (predicate.apply((d.getCategory()))) {
                     r.add(new Tag(c.ordinal(), d));
+                }
             } else {
-                if (predicate.apply(GlobalConfigurationCategory.get(Unclassified.class)))
+                if (predicate.apply(d.getCategory())) {
                     r.add(new Tag(0, d));
```
Comment (jglick) [inline core/src/main/java/hudson/Functions.java]: Why `0`? And if `getCategory` has been pushed up into `Descriptor`, why are we still checking `instanceof GlobalConfiguration`? The special-case call check of `Unclassified` was only necessary because that was the implied category of non-`GlobalConfigurationCategory` `Descriptor`s, but now these can actually declare their category. I would think the whole logic could be simplified to

```java
if (predicate.apply(d.getCategory())) {
    r.add(new Tag(c.ordinal(), d));
}
```

File: core/src/main/resources/hudson/model/Label/sidepanel.jelly (hunk 4 above)
Comment (jglick) [inline core/src/main/resources/hudson/model/Label/sidepanel.jelly]: Never mind:

> The logical approach seems to be changing regular Configure to use the gear icon, which frees the previously used tools icon for the tools configuration.

**Comment by jglick** [comment]
Needs some little fixes but :+1: overall, long overdue!

**Comment by jglick** [comment]
(and yes applying the new tabbed rendering to this page would be very welcome)

**Comment by aheritier** [comment]
a tab per tool  @jglick ? not sure how many sort of tools you can have but it might be annoying if it starts to scrolls horizontally (I didn't checked the behaviour of tabs when they don't fit in the screen width cc @tfennelly )

**Comment by jglick** [comment]
> a tab per tool

Well, per tool _type_, yes.

> I didn't checked the behaviour of tabs when they don't fit in the screen width

IIRC it wraps automatically.

File: core/src/main/java/jenkins/model/GlobalToolConfiguration.java (hunk 2 above)
Comment (jglick) [inline core/src/main/java/jenkins/model/GlobalToolConfiguration.java]: Will make that an :ant:. BTW if you add `@Restricted(NoExternalUse.class)` then we could freely move it later.

File: core/src/main/java/jenkins/model/GlobalConfigurationCategory.java (hunk 1 above)
Comment (jglick) [inline core/src/main/java/jenkins/model/GlobalConfigurationCategory.java]: :ant: or in a `jenkins.tools` package. I am thinking from the perspective of hypothetically splitting tool support to its own library plugin—then we would not want stuff like this inside more general types.

**Comment by jglick** [comment]
:bee: with reservations

**Comment by kohsuke** [comment]
According to Daniel, the remaining tasks are:

- [ ] update test cases
- [ ] rename package to accomodate Jesse's concern

**Comment by daniel-beck** [comment]
Superseded by #2140.

**Comment by amuniz** [comment]
This needs to be debugged because it's basically not working. For example, JDK installations are not persisted to disk, so they are all gone after a restart. Investigating...
